rust/main/submitter/src/payload_dispatcher/building_stage.rs

```diff
@@ -23,20 +27,113 @@ struct BuildingStage {
+            let payload = match self.queue.lock().await.pop_front() {
+                Some(payload) => payload,
+                None => {
+                    // wait for the next payload to arrive
+                    tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
+                    continue;
+                }
+            };
+
+            let payloads = vec![payload];
+            let txs = match self
+                .state
+                .adapter
+                .build_transactions(payloads.clone())
+                .await
+            {
+                Err(err) => {
+                    error!(
+                        ?err,
+                        ?payloads,
+                        "Error building transactions. Dropping payloads"
+                    );
+                    let details_for_payloads: Vec<_> =
+                        payloads.into_iter().map(|p| p.details().clone()).collect();
+                    self.drop_payloads(&details_for_payloads, DropReason::UnhandledError)
+                        .await;
+                    continue;
+                }
+                Ok(txs) => txs,
+            };
+
+            for tx in txs {
+                if let Err(err) = self.simulate_tx(&tx).await {
+                    error!(?err, "Error simulating transaction. Dropping transaction");
+                    // TODO: distinguish between network error and failed simulation
+                    self.drop_tx(&tx, DropReason::FailedSimulation).await;
+                    continue;
+                };
+                if let Err(err) = self.send_tx_to_inclusion_stage(tx.clone()).await {
+                    error!(?err, "Error storing transaction in the database");
+                    self.drop_tx(&tx, DropReason::UnhandledError).await;
+                    continue;
+                } else {
+                    self.store_tx(&tx).await;
                 }
-            } else {
-                // wait for the next payload to arrive
-                tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
             }
         }
     }
+
+    async fn drop_payloads(&self, details: &[PayloadDetails], reason: DropReason) {
+        for d in details {
+            if let Err(err) = self
+                .state
+                .payload_db
+                .set_payload_status(d.id(), PayloadStatus::Dropped(reason.clone()))
+                .await
+            {
+                error!(?err, "Error updating payload status to `dropped`");
+            }
+            warn!(?details, "Payload dropped from Building Stage");
+        }
+    }
+
+    async fn drop_tx(&self, tx: &Transaction, reason: DropReason) {
+        self.drop_payloads(tx.payload_details(), reason).await;
+    }
+
+    async fn store_tx(&self, tx: &Transaction) {
+        if let Err(err) = self.state.tx_db.store_transaction_by_id(tx).await {
+            error!(?err, "Error storing transaction in the database");
```

Comment:
Incorrect log message? 🤔

Reply:
I assume you were referring to the error from when `send_tx_to_inclusion_stage` fails rather? Updated it